Add script for finding feature defines

donny-dont opened the pull request:
#### b6841b91605ea8d2efd8dd163cf9560e5f55e215
<pre>
Add script for finding feature defines
<a href="https://bugs.webkit.org/show_bug.cgi?id=245290">https://bugs.webkit.org/show_bug.cgi?id=245290</a>

Reviewed by Darin Adler.

Adds a script for determining what feature defines (ENABLE, USE, HAVE)
are present in portions of WebKit code. As an example the script can
search all native code for all `ENABLE(flag)` and report back a full
listing. In addition it can also find differences between sections of
code which can be used to find stale build options.

* Tools/Scripts/find-feature-flags: Added.
* Tools/Scripts/webkitpy/featuredefines/__init__.py: Added.
* Tools/Scripts/webkitpy/featuredefines/matcher.py: Added.
* Tools/Scripts/webkitpy/featuredefines/matcher_unittest.py: Added.
* Tools/Scripts/webkitpy/featuredefines/search.py: Added.

Canonical link: <a href="https://commits.webkit.org/257596@main">https://commits.webkit.org/257596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/498fc656da279a68399d27f0b4aa5914b02e87d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99395 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8595 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108780 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/169021 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103392 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9158 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85907 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106703 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105152 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/33900 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/102916 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21810 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23327 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2368 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5226 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42802 "Passed tests") | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-9-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/find-feature-flags.py:
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python
```
Can we make this Python 3? And we usually drop the `.py` extension on scripts the user invokes directly.

donny-dont replied:
Yea sorry Windows doesn't do shebangs so its a force of habit to give actual extensions.

JonWBedard commented on Tools/Scripts/find-feature-flags.py:
```diff
@@ -0,0 +1,124 @@
+
+import webkitpy.featuredefines.search as search
+
+_DESCRIPTION = """
```
Do we need this variable if it's going to be empty?

donny-dont replied:
Oh sorry I just didn't write the documentation part yet.

JonWBedard commented on Tools/Scripts/webkitpy/featuredefines/search.py:
```diff
@@ -0,0 +1,152 @@
+class FeatureDefinesSearch(object):
+    """ Base class for searching for feature defines. """
+
+    ___metaclass___ = ABCMeta
```
I'm a bit surprised unit tests passed with this? This is a Python 3 specific idiom, if memory serves. It think it's fine for us to use this, but we should probably disable `webkitpy/featuredefines` in our Python 2 unit tests, and probably add an assert in the `__init__` or these `.py` files when they contain something Python 3 specific.

(as a note, there is intention to totally purge Python 2 from our codebase, but that's not high on our priority list at the moment)

donny-dont replied:
There aren't any associated tests for this file so maybe that's why?

donny-dont replied:
Are there examples on how to mark something as Python 3 only in webkitpy?

JonWBedard replied:
> Are there examples on how to mark something as Python 3 only in webkitpy?

There are not, but the gist of it would be to add `tester.skip` in `Tools/Scripts/webkitpy/test/main.py` and then do something like:
```
if sys.version_info < (3, 0):
    raise NotImplementedError
```

donny-dont replied:
Actually looks like itd be ok https://python.readthedocs.io/en/v2.7.2/library/abc.html

lauromoura commented on Tools/Scripts/webkitpy/featuredefines/matcher.py:
```diff
@@ -0,0 +1,108 @@
+
+def cmake_options_matcher(macro='ENABLE'):
+    """ Matches WEBKIT_OPTION_DEFINE(MACRO_FLAG) """
+    pattern = r'WEBKIT_OPTION_DEFINE\((?P<flag>' + macro + r'_[0-9A-Z_]+) "(?P<description>.*)" PRIVATE (?P<value>.*)\)'
```
What about `PUBLIC` options? (e.g. `ENABLE_X11_TARGET`, `USE_SOUP2`, ` ENABLE_STATIC_JSC`, etc).

donny-dont replied:
Good catch! All the `WEBKIT_OPTION_DEFINE` in `WebKitFeatures.cmake` are `PRIVATE` so that's why I was doing it.

darinadler reviewed (approved):
This script looks useful. I think it’s fine to land as is, or you could take the @JonWBedard suggestions. If we find it’s rarely used we can remove it.

donny-dont committed:
Add script for finding feature defines
https://bugs.webkit.org/show_bug.cgi?id=245290

Reviewed by Darin Adler.

Adds a script for determining what feature defines (ENABLE, USE, HAVE)
are present in portions of WebKit code. As an example the script can
search all native code for all `ENABLE(flag)` and report back a full
listing. In addition it can also find differences between sections of
code which can be used to find stale build options.

* Tools/Scripts/find-feature-flags: Added.
* Tools/Scripts/webkitpy/featuredefines/__init__.py: Added.
* Tools/Scripts/webkitpy/featuredefines/matcher.py: Added.
* Tools/Scripts/webkitpy/featuredefines/matcher_unittest.py: Added.
* Tools/Scripts/webkitpy/featuredefines/search.py: Added.

Canonical link: https://commits.webkit.org/257596@main
